generators/chipyard/src/main/scala/ChipTop.scala

```diff
@@ -0,0 +1,312 @@
+package chipyard.chiptop
+
+import chisel3._
+import chipyard.{DigitalTop, System, SystemModule}
+
+import scala.collection.mutable.{ArrayBuffer}
+
+import chisel3.experimental.{Analog}
+import freechips.rocketchip.config.{Parameters, Field}
+import freechips.rocketchip.diplomacy.{LazyModule}
+import freechips.rocketchip.devices.debug.{DebugIO, PSDIO, HasPeripheryDebugModuleImp}
+import freechips.rocketchip.subsystem.{HasExtInterruptsModuleImp, CanHaveMasterAXI4MemPortModuleImp}
+import sifive.blocks.devices.gpio.{GPIOPortIO, HasPeripheryGPIOModuleImp}
+import sifive.blocks.devices.uart.{UARTPortIO, HasPeripheryUARTModuleImp}
+import testchipip.{BlockDeviceIO, CanHavePeripheryBlockDeviceModuleImp, SerialIO, CanHavePeripherySerialModuleImp}
+import icenet.{NICIOvonly, CanHavePeripheryIceNICModuleImp}
+import tracegen.{HasTraceGenTilesModuleImp}
+import chipyard.config.ConfigValName._
+
+import barstools.iocell.chisel._
+
+case object BuildSystem extends Field[Parameters => SystemModule[System]]((p: Parameters) => Module(LazyModule(new DigitalTop()(p)).suggestName("system").module))
+
+abstract class BaseChipTop(val useIOCells: Boolean = true)(implicit val p: Parameters) extends RawModule {
+
+  val iocells = ArrayBuffer.empty[IOCell]
+
+  val systemClock = Wire(Input(Clock()))
+  val systemReset = Wire(Input(Bool()))
+
+  val system = withClockAndReset(systemClock, systemReset) { p(BuildSystem)(p) }
+
+  def connectReset(r: Bool): Unit
+
+}
+
+trait HasChipTopSimpleClockAndReset { this: BaseChipTop =>
+  implicit val p: Parameters
+
+  val (clock, systemClockIO) = if (useIOCells) {
+    IOCell.generateIOFromSignal(systemClock, Some("iocell_clock"))
+  } else {
+    val port = IO(Input(Clock()))
+    systemClock := port
+    (port, Seq())
+  }
+
+  // TODO how to handle async vs sync reset?
+  val (reset, systemResetIO) = if (useIOCells) {
+    IOCell.generateIOFromSignal(systemReset, Some("iocell_reset"))
+  } else {
+    val port = IO(Input(Bool()))
+    systemReset := port
+    (port, Seq())
+  }
+
+  iocells ++= systemClockIO
+  iocells ++= systemResetIO
+
+  def connectReset(r: Bool) { reset := r }
```

Comment:
Why is this `reset := r`, instead of `systemReset := r`?

Reply:
`systemReset` is the internal wire, which is already connected to the `reset` IO. This method is needed because `reset` is not a member of the base class.

Reply:
We probably need this for clock, too...

Reply:
Never mind- we only need it for reset so we can override the default reset connection when we mix in the `Debug` trait

Reply:
This was removed in a later commit.